azurelinuxagent/ga/update.py

```diff
@@ -986,6 +1009,55 @@ def _ensure_firewall_rules_persisted(dst_ip):
             message=msg,
             log_event=False)
 
+    def __upgrade_agent_if_permitted(self):
```

Comment:
"permitted" is not the right term... that makes me think that this is related to AutoUpdate being enabled or not, when it is actually related to the period of the update

Reply:
I think keeping __upgrade_agent_if_permitted and _check_and_download_agent_if_upgrade_available as separate operations makes the logic harder to understand. Wouldn't it be simpler to do the download and upgrade within the same operation?

Currently we have that issue where a bad error.json raises an exception. by calling get_latest_agent() within __upgrade_agent_if_permitted() I think that condition would prevent the update from happening.

Reply:
> "permitted" is not the right term... that makes me think that this is related to AutoUpdate being enabled or not, when it is actually related to the period of the update

Fair, I'll rename the function to `__upgrade_agent_if_time_elapsed`

> I think keeping __upgrade_agent_if_permitted and _check_and_download_agent_if_upgrade_available as separate operations makes the logic harder to understand. Wouldn't it be simpler to do the download and upgrade within the same operation?

I purposefully kept it separate, like I dont want the upgrading to be dependent on the download because that would add atleast an overhead of 1 hour (i.e. we'll start our counter after we download which is every one hour). The idea behind keeping it separate is to kind of trickle the upgrades as and when they come because they would be dependent on the service start time. I can explain more over a call if you would prefer that.

> Currently we have that issue where a bad error.json raises an exception. by calling get_latest_agent() within __upgrade_agent_if_permitted() I think that condition would prevent the update from happening.

Good point, though I feel that bug will block the updates even now. I think we can handle that bug separately as part of the task to phase out blacklisting logic.

Reply:
It's OK as is, I think the implementation can be further simplified but you have a fair point that this is just temporary code while the Agent versioning get implemented, so no issue here.